fix:logbucket sink in same project

felipecrescencio-cit opened the pull request:
Fix for issue 116.
PR changes:
- New variable `grant_write_permission_on_bkt` on logbucket destination module that enables the user to not grant logbucket write permission, that is the use case when sink and logbucket are in the same project - [reference](https://cloud.google.com/logging/docs/export/configure_export_v2#dest-auth:~:text=If%20you%27re%20using%20a%20sink%20to%20route%20logs%20between%20Logging%20buckets%20in%20the%20same%20Cloud%20project%2C%20no%20new%20service%20account%20is%20created%3B%20the%20sink%20works%20without%20the%20unique%20writer%20identity)
- Example of usage logbucket in projects
- Integrated test of logbucket with projects

felipecrescencio-cit committed:
Fix Copyright year

felipecrescencio-cit committed:
New variable to deal with sink and logbucket in same project

felipecrescencio-cit committed:
When sink and logbucket in same project no new service account is created

felipecrescencio-cit committed:
New examples for logbucket at project-level

felipecrescencio-cit committed:
Changed name of new variable to a meaningful name

felipecrescencio-cit committed:
Created example of logbucket with project usage

felipecrescencio-cit committed:
Changed name of bucket due to colisions during integrated tests

felipecrescencio-cit committed:
Setup a new project for logbucket project testing in same and other projects

felipecrescencio-cit committed:
Integrated tests to Logbucket Project example

felipecrescencio-cit committed:
Revert mischanged copyright year

felipecrescencio-cit committed:
Shorten variable names

daniel-cit commented on examples/logbucket/project/README.md:
```diff
@@ -0,0 +1,30 @@
+# Log Export: Log Bucket destination at Folder level
```
```suggestion
# Log Export: Log Bucket destination at Project level
```

daniel-cit commented on examples/logbucket/project/README.md:
```diff
@@ -0,0 +1,30 @@
+# Log Export: Log Bucket destination at Folder level
+
+These examples configures a project-level log sink that feeds a logging log bucket destination with log bucket and log sink in the same project and in separated projects.
```
```suggestion
These examples configures a project-level log sink that feeds a logging log bucket destination with log bucket and log sink in the same project or in separated projects.
```

daniel-cit commented on modules/logbucket/outputs.tf:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright 2019 Google LLC
+ * Copyright 2022 Google LLC
```
you don't need to change the date of existing files

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+
+	const logApiFdqm = "logging.googleapis.com"
+	const RETENTION_DAYS = 20
+	const ROLE_BUCKET_WRITER string = "roles/logging.bucketWriter"
```
this could be inlined

felipecrescencio-cit replied:
Done

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+func TestLogBucketProjectModule(t *testing.T) {
+
+	const logApiFdqm = "logging.googleapis.com"
+	const RETENTION_DAYS = 20
```
this cloud be inlined

daniel-cit replied:
RETENTION_DAYS  is not been used and it is not 20 days

felipecrescencio-cit replied:
Created constant defaultRetentionDays and kept both examples with this default for user better understanding code and example

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+		logBucketDetails := gcloud.Runf(t, fmt.Sprintf("logging buckets describe %s --location=%s --project=%s", bktName, "global", projId))
+
+		// assert log bucket name, retention days & location
+		assert.Equal(sinkDest[len(logApiFdqm)+1:], logBucketDetails.Get("name").String(), "log bucket name should match")
```
what is this code doing?
```
sinkDest[len(logApiFdqm)+1:]
```

daniel-cit replied:
I think it would be more clear if this was
```
assert.Contains(sinkDest, logBucketDetails.Get("name").String() ...
```

felipecrescencio-cit replied:
Changed assert bucket name to remove substring and compare to a built string fully qualified name

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+		//*******************************
+		// Bucket on other project test *
+		//*******************************
+		projId := bpt.GetStringOutput("log_bucket_project")
+		bktName := bpt.GetStringOutput("log_bucket_name")
+		bktDestProjId := bpt.GetStringOutput("log_bkt_same_proj")
+		sinkProjId := bpt.GetStringOutput("log_sink_project_id")
+		sinkDest := bpt.GetStringOutput("log_sink_destination_uri")
+		sinkName := bpt.GetStringOutput("log_sink_resource_name")
+		sinkWriterIdentity := bpt.GetStringOutput("log_sink_writer_identity")
```
we don't create all the variables in advance at the beginning of the code, they are usually created just before first usage.
Same for the other set of variables

felipecrescencio-cit replied:
Fixed this by combining similar asserts to tabular

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+
+		// assert log bucket name, retention days & location
+		assert.Equal(sameProjSinkDest[len(logApiFdqm)+1:], sameProjBktDetails.Get("name").String(), "log bucket name should match")
+		assert.Equal(int64(RETENTION_DAYS), sameProjBktDetails.Get("retentionDays").Int(), "retention days should match")
```
two diferente values?
```
assert.Equal(int64(30), logBucketDetails.Get("retentionDays").Int(), "retention days should match")
```

felipecrescencio-cit replied:
Created constant defaultRetentionDays and kept both examples with this default for user better understanding code and example

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+		bpt.DefaultVerify(assert)
+
+		//*******************************
+		// Bucket on other project test *
```
is it possible to combine these two test in one tabular test like
https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/blob/master/infra/blueprint-test/test/tfvet_test.go#L38

felipecrescencio-cit replied:
Created a `for` with a struct to handle similar asserts based on the code metioned

felipecrescencio-cit committed:
Fix logbucket/project example README.md

Co-authored-by: Daniel Andrade <dandrade@ciandt.com>

felipecrescencio-cit committed:
Fix logbucket/project example README.md

Co-authored-by: Daniel Andrade <dandrade@ciandt.com>

felipecrescencio-cit committed:
Revert mischanged copyright year

felipecrescencio-cit committed:
Inline constants

felipecrescencio-cit committed:
Changed variables creation place to right before usage

felipecrescencio-cit committed:
Remove retention_days to keep it default user better understanding example

felipecrescencio-cit committed:
-Created constant defaultRetentionDays and kept both examples with this default for user better understanding code and example
-Changed assert bucket name to remove substring and compare to a built string fully qualified name
-Combined similar asserts to tabular

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+				sinkDest:           bpt.GetStringOutput("log_sink_dest_uri_same_proj"),
+				sinkProjId:         bpt.GetStringOutput("log_sink_id_same_proj"),
+				sinkName:           bpt.GetStringOutput("log_sink_resource_name_same_proj"),
+				sinkWriterIdentity: "",
```
```suggestion
```
default value for a variable of type string is ""

felipecrescencio-cit replied:
I removed and left a comment explaning that in this example writerIdentity will be empty

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+			sinkDest           string
+			sinkProjId         string
+			sinkName           string
+			sinkWriterIdentity string
```
```suggestion
			writerIdentity string
```
you don't need to prefix with something else because there is only one writerIdentity

felipecrescencio-cit replied:
Prefix removed

daniel-cit commented on test/integration/logbucket-project/logbucket_project_test.go:
```diff
@@ -0,0 +1,96 @@
+			// assert log sink name, destination & filter
+			assert.Equal(tc.sinkDest, logSinkDetails.Get("destination").String(), "log sink destination should match")
+			assert.Equal("resource.type = gce_instance", logSinkDetails.Get("filter").String(), "log sink filter should match")
+			assert.Equal(tc.sinkWriterIdentity, logSinkDetails.Get("writerIdentity").String(), "log sink writerIdentity should not be empty")
```
"log sink writerIdentity should not be empty" ??
it is empty in the second case.
maybe "should have excepted  log sink writerIdentity" ?

felipecrescencio-cit replied:
I keep "log sink writerIdentity should match" and wrote a comment on the empty writerIdentity example:
```
sinkName:   bpt.GetStringOutput("log_sink_resource_name_same_proj"),
// writerIdentity: As sink and bucket are in same project no service account is needed and writerIdentity is empty
```

felipecrescencio-cit commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Problem having log-export and logbucket destination on same project](https://trello.com/c/U8ktQg89/367-problem-having-log-export-and-logbucket-destination-on-same-project)

felipecrescencio-cit committed:
Removed because empty is string default value

Co-authored-by: Daniel Andrade <dandrade@ciandt.com>

felipecrescencio-cit committed:
Removed sink prefix because there is only one writerIdentity

Co-authored-by: Daniel Andrade <dandrade@ciandt.com>

felipecrescencio-cit committed:
Fix assert writerIdentity message for examples

felipecrescencio-cit commented:
Internal review finished.
PR 118 opened in the official repo.
